<a href="https://colab.research.google.com/github/Kari-2005/carisurg-portfolio/blob/main/Assignment_2/Assignment_2_Cleaned_MAP_Column_Personal_Attempt_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# CARISURG

# CARISURG WEEK 0 - ASSIGNMENT 1 - Data Cleaning (Gender Column)
## Submitted by Ansarah Mohammed

In [2]:
# Mounting my Google Drive
from google.colab import drive
drive.mount('/content/drive')
print("Drive mounted successfully!")

Mounted at /content/drive
Drive mounted successfully!


In [3]:
# Confirming the python version
import sys
print(f"Python version: {sys.version}")

Python version: 3.12.13 (main, Mar  4 2026, 09:23:07) [GCC 11.4.0]


In [4]:
# Import the libraries needed for data cleaning
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

print("Libraries are successfully imported for data cleaning!")

Libraries are successfully imported for data cleaning!


In [5]:
# Load the dataset
FILE_PATH = "/content/drive/MyDrive/CARISURG/Week0/EmergencyTriageDataset_Reduced_Dirty.csv"

df_raw = pd.read_csv(FILE_PATH)
print(f"Dataset Loaded: {df_raw.shape[0]} rows x {df_raw.shape[1]} columns")
print(f"Columns: {list(df_raw.columns)}")

Dataset Loaded: 2205 rows x 11 columns
Columns: ['ID', 'Age', 'Gender', 'GCS', 'SBP', 'DBP', 'MAP', 'pulse', 'Temp', 'RR', 'Fio2']


In [6]:
# Examining Rows
df_raw.head(10)

,ID,Age,Gender,GCS,SBP,DBP,MAP,pulse,Temp,RR,Fio2
0,1,34,0,15.0,93,67.0,75.67,128.0,36.8,14.0,21.0
1,2,20,Male,15.0,130,90.0,103.33,80.0,37.0,16.0,21.0
2,3,77,Female,14.0,163,105.0,124.33,92.0,36.8,18.0,21.0
3,4,23,0,8.0,100,60.0,73.33,100.0,37.0,12.0,100.0
4,5,86,FEMALE,15.0,150,90.0,110.00,85.0,37.0,19.0,21.0
5,6,42,Male,15.0,100,60.0,73.33,99.0,37.0,20.0,21.0
6,7,75,Female,15.0,120,80.0,93.33,99.0,37.0,25.0,21.0
7,8,25,Male,15.0,100,50.0,66.67,85.0,37.0,25.0,21.0
8,9,67,0,15.0,110,70.0,83.33,78.0,37.0,16.0,21.0
9,11,82,Female,15.0,153,82.0,105.67,130.0,37.0,19.0,21.0


In [7]:
# Looking at what types panda thinks each column is
print(df_raw.dtypes)

ID          int64
Age         int64
Gender     object
GCS        object
SBP        object
DBP       float64
MAP       float64
pulse      object
Temp       object
RR        float64
Fio2      float64
dtype: object


In [8]:
# Explore the values in the Gender Column
print("Unique values in Gender:")
print(df_raw['Gender'].unique())
print(f"\nTotal unique values: {df_raw['Gender'].nunique()}")

Unique values in Gender:
['0' 'Male' 'Female' 'FEMALE' '1' 'MALE']

Total unique values: 6


In [9]:
# Quantity of each unique value stored in Gender Column
print("Gender value quantity:")
print(df_raw['Gender'].value_counts())

Gender value quantity:
Gender
1         422
MALE      379
Male      375
FEMALE    366
Female    340
0         323
Name: count, dtype: int64


In [10]:
#In this design, gender is converted into numerical values:
# 1 - Male, 0 = Female. (Non-binary = 2 for future consideration)
# The current dataset only contains Male/ Female values but non-binary values will be included for future handling of genders.
# For a more efficient/ alternate way we can first convert all values to lowercase

gender_map = {
    'male': 1,
    'female': 0,
    '1': 1,
    '0': 0,
    # adding future considerations (m/f/non-binary)
    'm': 1,
    'f': 0,
    'non-binary': 2,
    'nonbinary': 2,
    'nb': 2,
    '2': 2
}

In [11]:
# Cleaning the column (for efficient mapping we convert each to lowercase before we map
df_raw['Gender_Cleaned'] = (df_raw['Gender'].astype(str).str.strip().str.lower().map(gender_map))

In [12]:
# Verifying the result after cleaning
print("After Cleaning:")
print(df_raw['Gender_Cleaned'].value_counts())
print(f"\nAny NaN values? {df_raw['Gender_Cleaned'].isnull().sum()}")


After Cleaning:
Gender_Cleaned
1    1176
0    1029
Name: count, dtype: int64

Any NaN values? 0


In [13]:
# Replacing the original dirty column with the new cleaned data
if 'Gender' in df_raw.columns: df_raw = df_raw.drop(columns=['Gender'])

df_raw = df_raw.rename(columns={'Gender_Cleaned': 'Gender'})

# Confirm the result of the new columns
print("Column 'Gender' after cleaning:")
print(df_raw['Gender'].value_counts())
print(f"Data type: {df_raw['Gender'].dtype}")

df_raw.head()

Column 'Gender' after cleaning:
Gender
1    1176
0    1029
Name: count, dtype: int64
Data type: int64


,ID,Age,GCS,SBP,DBP,MAP,pulse,Temp,RR,Fio2,Gender
0,1,34,15.0,93,67.0,75.67,128.0,36.8,14.0,21.0,0
1,2,20,15.0,130,90.0,103.33,80.0,37.0,16.0,21.0,1
2,3,77,14.0,163,105.0,124.33,92.0,36.8,18.0,21.0,0
3,4,23,8.0,100,60.0,73.33,100.0,37.0,12.0,100.0,0
4,5,86,15.0,150,90.0,110.00,85.0,37.0,19.0,21.0,0


# CARISURG WEEK 0 - ASSIGNMENT 2 - Data Cleaning (MAP Column) Personal Attempt
## Submitted by Ansarah Mohammed

In [14]:
# The five most useful inspection commands — run them in order, every time
df = pd.read_csv(FILE_PATH)
print("=== head() — first 5 rows ===")
print(df.head())

=== head() — first 5 rows ===
   ID  Age  Gender   GCS  SBP    DBP     MAP  pulse  Temp    RR   Fio2
0   1   34       0  15.0   93   67.0   75.67  128.0  36.8  14.0   21.0
1   2   20    Male  15.0  130   90.0  103.33   80.0  37.0  16.0   21.0
2   3   77  Female  14.0  163  105.0  124.33   92.0  36.8  18.0   21.0
3   4   23       0   8.0  100   60.0   73.33  100.0  37.0  12.0  100.0
4   5   86  FEMALE  15.0  150   90.0  110.00   85.0  37.0  19.0   21.0


In [15]:
print("=== tail() — last 5 rows (catches truncation issues) ===")
print(df.tail())

=== tail() — last 5 rows (catches truncation issues) ===
        ID  Age  Gender   GCS  SBP    DBP     MAP  pulse  Temp    RR  Fio2
2200  2380   65       1  14.0  170   50.0   90.00   65.0  37.0  22.0  21.0
2201  2381   84  FEMALE  15.0  180  120.0  140.00   85.0  36.1  18.0  21.0
2202  2382   78  FEMALE  15.0  132   78.0   96.00  100.0  37.9  16.0  21.0
2203  2383   65    Male  15.0  140   80.0  100.00   85.0  36.4  16.0  21.0
2204  2384   78  FEMALE  15.0  145   90.0  108.33   75.0  37.6  20.0  21.0


In [16]:
print("=== info() — column types and non-null counts ===")
df.info()

=== info() — column types and non-null counts ===
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2205 entries, 0 to 2204
Data columns (total 11 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      2205 non-null   int64  
 1   Age     2205 non-null   int64  
 2   Gender  2205 non-null   object 
 3   GCS     2183 non-null   object 
 4   SBP     2205 non-null   object 
 5   DBP     2183 non-null   float64
 6   MAP     2183 non-null   float64
 7   pulse   2183 non-null   object 
 8   Temp    2183 non-null   object 
 9   RR      2183 non-null   float64
 10  Fio2    2183 non-null   float64
dtypes: float64(4), int64(2), object(5)
memory usage: 189.6+ KB


In [17]:
print("=== describe() — statistics for numeric columns only ===")
print(df.describe())

=== describe() — statistics for numeric columns only ===
                ID          Age          DBP          MAP           RR  \
count  2205.000000  2205.000000  2183.000000  2183.000000  2183.000000   
mean   1154.987755    61.829478    77.479615    93.912277    20.262254   
std     677.167364    18.485363    16.713289    19.006296     5.742333   
min       1.000000    18.000000    30.000000    39.330000    12.000000   
25%     577.000000    50.000000    70.000000    82.500000    17.000000   
50%    1135.000000    64.000000    78.000000    93.330000    18.000000   
75%    1703.000000    77.000000    87.000000   103.330000    21.000000   
max    2384.000000    98.000000   173.000000   185.000000    50.000000   

              Fio2  
count  2183.000000  
mean     25.019698  
std      10.144288  
min      21.000000  
25%      21.000000  
50%      21.000000  
75%      21.000000  
max     100.000000  


In [18]:
print("=== shape — (rows, columns) ===")
print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")

=== shape — (rows, columns) ===
Rows: 2205, Columns: 11



**Clinical Reference Table:**

| Column | Clinical Meaning | Valid Range | Unit |
|--------|-----------------|-------------|------|
| `Age` | Patient age | 0 – 120 | years |
| `Gender` | Biological sex | 0 (Female) or 1 (Male) | encoded |
| `GCS` | Glasgow Coma Scale — consciousness level | 3 – 15 | score |
| `SBP` | Systolic Blood Pressure | 50 – 250 | mmHg |
| `DBP` | Diastolic Blood Pressure | 30 – 150 | mmHg |
| `MAP` | Mean Arterial Pressure — organ perfusion indicator | 40 – 180 | mmHg |
| `pulse` | Heart rate | 20 – 250 | bpm |
| `Temp` | Body temperature | 32.0 – 43.0 | °C |
| `RR` | Respiratory rate | 5 – 60 | breaths/min |
| `Fio2` | Fraction of Inspired Oxygen | 21 – 100 | % |

**Key insight:** The valid range is determined by *physiology*, not by the data.
A `GCS` of 99 is not an outlier — it is an error. They are different things.


 **MAP Cleaning — SBP and DBP:**

Before calculating Mean Arterial Pressure (MAP), I decided to clean the SBP and DBP columns first because MAP is directly calculated from these two values. SBP represents the systolic blood pressure, while DBP represents the diastolic blood pressure. If either of these columns contains missing values, text values, or unrealistic readings, the MAP result may become inaccurate.

Since MAP is calculated using the formula:

MAP = DBP + (SBP - DBP) / 3

both SBP and DBP need to be valid before applying the calculation. For example, if SBP or DBP has an out-of-range value, that error would carry over into the MAP column. Because of this, I cleaned the blood pressure values first by checking their data types, identifying missing values, and looking for values outside a reasonable medical range. This helped ensure that the final MAP values were based on reliable blood pressure readings.

In [19]:
# SBP = Systolic Blood Pressure Cleaning.

In [20]:
# Valid range: 50–250 mmHg.
# First we inspect

print("SBP unique values (sample):", df['SBP'].unique()[:15])
print("SBP dtype:", df['SBP'].dtype)

SBP unique values (sample): ['93' '130' '163' '100' '150' '120' '110' '153' '152' '186' '107' '80'
 '75' '58' '140']
SBP dtype: object


In [21]:
# Convert to non-numerical entries to numerical (becomes NaN)
df['SBP'] = pd.to_numeric(df['SBP'], errors='coerce')
print("After Numerical conversion:")
print(df['SBP'].describe())

After Numerical conversion:
count    2183.000000
mean      128.388456
std        43.881583
min        30.000000
25%       110.000000
50%       125.000000
75%       140.000000
max       500.000000
Name: SBP, dtype: float64


In [22]:
# Identify any values that are out-of-range
invalid_sbp = df[(df['SBP'] < 50) | (df['SBP'] > 250)]
print(f"Out-of-range SBP rows: {len(invalid_sbp)}")
print(invalid_sbp[['ID', 'SBP']].head())

Out-of-range SBP rows: 44
      ID    SBP
29    32   30.0
111  123  500.0
168  183  500.0
231  247   30.0
275  291  500.0


In [23]:
# Replace any out-of-range values with NaN, then impute with median.
# The median was used to impute SBP because it is not strongly affected by outliers.
# This makes it a more reliable replacement value for missing or invalid blood pressure
# readings than the mean.

df.loc[(df['SBP'] < 50) | (df['SBP'] > 250), 'SBP'] = np.nan
sbp_median = df['SBP'].median()
df['SBP'] = df['SBP'].fillna(sbp_median)

print(f"SBP after cleaning: min={df['SBP'].min()}, max={df['SBP'].max()}, NaNs={df['SBP'].isnull().sum()}")

SBP after cleaning: min=55.0, max=250.0, NaNs=0


In [24]:
# DSP = Diastolic Blood Pressure

In [25]:
# Valid Range: 30-150 mmHg
# First we inspect

print("DBP dtype:", df["DBP"].dtype)
print("DBP unique values sample:", df["DBP"].unique()[:20])
print("DBP missing values before cleaning:", df["DBP"].isnull().sum())

DBP dtype: float64
DBP unique values sample: [ 67.  90. 105.  60.  80.  50.  70.  82.  89. 100.  84.  40.  61.  32.
  53.  63.  nan  76.  75.  48.]
DBP missing values before cleaning: 22


In [26]:
# Converting DBP to numeric
df["DBP"] = pd.to_numeric(df["DBP"], errors="coerce")

print("\nDBP dtype after conversion:", df["DBP"].dtype)
print("DBP missing values after conversion:", df["DBP"].isnull().sum())


DBP dtype after conversion: float64
DBP missing values after conversion: 22


In [27]:
# Flagging invalid DBP values
# Valid DBP range: 30-150 mmHg
invalid_dbp = df[(df["DBP"] < 30) | (df["DBP"] > 150)]

print(f"\nDBP values outside valid range (30-150): {len(invalid_dbp)} rows")
print(invalid_dbp[["ID", "DBP"]].head(10))


DBP values outside valid range (30-150): 3 rows
        ID    DBP
375    395  173.0
951    981  160.0
1468  1510  160.0


In [28]:
# Replacing invalid DBP values with NaN
df.loc[(df["DBP"] < 30) | (df["DBP"] > 150), "DBP"] = np.nan

print("\nDBP missing values after range filtering:", df["DBP"].isnull().sum())


DBP missing values after range filtering: 25


In [29]:
# Imputing missing DBP values using the median
# Median is used because blood pressure values can contain outliers,
# and the median is less affected by extreme values than the mean.

dbp_median = df["DBP"].median()

df["DBP"] = df["DBP"].fillna(dbp_median)

print(f"\nMedian DBP used for imputation: {dbp_median}")
print("DBP missing values after median imputation:", df["DBP"].isnull().sum())

print("\nDBP summary after imputation:")
print(df["DBP"].describe())

print("DBP missing values after range filtering:", df["DBP"].isnull().sum())


Median DBP used for imputation: 78.0
DBP missing values after median imputation: 0

DBP summary after imputation:
count    2205.000000
mean       77.367347
std        16.316160
min        30.000000
25%        70.000000
50%        78.000000
75%        87.000000
max       150.000000
Name: DBP, dtype: float64
DBP missing values after range filtering: 0


In [30]:
# MAP- Mean Arterial Pressure

In [31]:
# Valid Range: 40-180 mmHg
# Calculating MAP

df["MAP"] = (df["SBP"] + 2 * df["DBP"]) / 3

In [32]:
# After calculating MAP, we inspect the new values

print("MAP missing values:", df["MAP"].isnull().sum())
print("MAP range:", df["MAP"].min(), "-", df["MAP"].max())
print(df[["SBP", "DBP", "MAP"]].head())

MAP missing values: 0
MAP range: 39.333333333333336 - 173.33333333333334
     SBP    DBP         MAP
0   93.0   67.0   75.666667
1  130.0   90.0  103.333333
2  163.0  105.0  124.333333
3  100.0   60.0   73.333333
4  150.0   90.0  110.000000


In [33]:
# Check for invalid MAP values

invalid_map = df[(df["MAP"] < 40) | (df["MAP"] > 180)]

print("MAP values outside valid range:", len(invalid_map))
print(invalid_map[["SBP", "DBP", "MAP"]].head())

MAP values outside valid range: 1
       SBP   DBP        MAP
1710  58.0  30.0  39.333333


In [34]:
# Flag MAP values that may be clinically low but still possible
# <60 and >180 is alarming as this means that there is a serious medical problem.

df["MAP_flag"] = "Normal/Acceptable"

df.loc[df["MAP"] < 60, "MAP_flag"] = "Low MAP - review"
df.loc[df["MAP"] > 180, "MAP_flag"] = "High MAP - review"

# Display rows where MAP needs review
print(df[df["MAP_flag"] != "Normal/Acceptable"][["SBP", "DBP", "MAP", "MAP_flag"]])

       SBP   DBP        MAP          MAP_flag
20    80.0  40.0  53.333333  Low MAP - review
26    58.0  32.0  40.666667  Low MAP - review
114   76.0  51.0  59.333333  Low MAP - review
140   69.0  45.0  53.000000  Low MAP - review
147   74.0  39.0  50.666667  Low MAP - review
157   80.0  40.0  53.333333  Low MAP - review
234   70.0  50.0  56.666667  Low MAP - review
250   55.0  40.0  45.000000  Low MAP - review
295   70.0  50.0  56.666667  Low MAP - review
296   70.0  40.0  50.000000  Low MAP - review
310   60.0  50.0  53.333333  Low MAP - review
315   80.0  48.0  58.666667  Low MAP - review
337   75.0  50.0  58.333333  Low MAP - review
352   73.0  40.0  51.000000  Low MAP - review
367   68.0  50.0  56.000000  Low MAP - review
386   80.0  40.0  53.333333  Low MAP - review
429   84.0  47.0  59.333333  Low MAP - review
430   81.0  47.0  58.333333  Low MAP - review
443   74.0  47.0  56.000000  Low MAP - review
451   90.0  40.0  56.666667  Low MAP - review
482   82.0  44.0  56.666667  Low M

In [35]:
# Final validation check for SBP, DBP, and MAP

# A list of columns to check, along with their accepted minimum and maximum values
checks = [
    ("SBP", 50, 250),    # SBP should be between 50 and 250 mmHg
    ("DBP", 30, 150),    # DBP should be between 30 and 150 mmHg
    ("MAP", 40, 180)     # MAP should be between 40 and 180 mmHg
]

# Validation results
print("Final Validation Check")
print("----------------------")

# Loop through each column and its valid range
for col, lo, hi in checks:

    # Count how many missing values are in the column
    n_missing = df[col].isnull().sum()

    # Count how many values are outside the accepted range
    n_invalid = ((df[col] < lo) | (df[col] > hi)).sum()

    # Find the minimum and maximum values in the column
    min_val = df[col].min()
    max_val = df[col].max()

    # If there are no missing or invalid values, the column passes the check
    if n_missing == 0 and n_invalid == 0:
        status = "PASS"
    else:
        status = "CHECK"

    # Print the validation results for the current column
    print(f"{status} {col}:")
    print(f"  Range: {min_val:.1f} - {max_val:.1f}")
    print(f"  Missing values: {n_missing}")
    print(f"  Out-of-range values: {n_invalid}")
    print()

Final Validation Check
----------------------
PASS SBP:
  Range: 55.0 - 250.0
  Missing values: 0
  Out-of-range values: 0

PASS DBP:
  Range: 30.0 - 150.0
  Missing values: 0
  Out-of-range values: 0

CHECK MAP:
  Range: 39.3 - 173.3
  Missing values: 0
  Out-of-range values: 1



In [36]:
# Display final cleaned blood pressure columns

print(df[["SBP", "DBP", "MAP"]].head(15))

      SBP    DBP         MAP
0    93.0   67.0   75.666667
1   130.0   90.0  103.333333
2   163.0  105.0  124.333333
3   100.0   60.0   73.333333
4   150.0   90.0  110.000000
5   100.0   60.0   73.333333
6   120.0   80.0   93.333333
7   100.0   50.0   66.666667
8   110.0   70.0   83.333333
9   153.0   82.0  105.666667
10  150.0   90.0  110.000000
11  110.0   90.0   96.666667
12  120.0   80.0   93.333333
13  152.0   89.0  110.000000
14  130.0   70.0   90.000000


In [37]:
OUTPUT_PATH = 'Emergency_Triage_Dataset_Day1&2_Cleaned_AnsarahM.csv'
df.to_csv(OUTPUT_PATH, index=False)
print(f'Saved → {OUTPUT_PATH}')

Saved → Emergency_Triage_Dataset_Day1&2_Cleaned_AnsarahM.csv
